src/neo/Ledger/Blockchain.cs

```diff
@@ -495,15 +497,20 @@ private void Persist(Block block)
                         Transaction = tx
                     };
 
-                    snapshot.Transactions.Add(tx.Hash, state);
+                    clonedSnapshot.Transactions.Add(tx.Hash, state);
+                    clonedSnapshot.Transactions.Commit();
 
-                    using (ApplicationEngine engine = new ApplicationEngine(TriggerType.Application, tx, snapshot.Clone(), tx.SystemFee))
+                    using (ApplicationEngine engine = new ApplicationEngine(TriggerType.Application, tx, clonedSnapshot, tx.SystemFee))
                     {
                         engine.LoadScript(tx.Script);
                         state.VMState = engine.Execute();
```

Comment:
The commit in line 501 have no sense, because the state it's changed in this line.

Reply:
It doesn't matter, after `clonedSnapshot.Transactions.Commit()` , tx would be added in `snapshot` with `Add` flag. Therefore, this line change will be persisted in `snapshot`.

Reply:
Could you test it with a FAULT transaction?

Reply:
@Qiao-Jin need your help.

Reply:
Hi @shargon, I tested this case (fault transaction) in previously uploaded UT and it works, which can prove that fault transactions can be handled correctly in my local env. However this newly added UT brings problems to UT_Consensus, as Consensus.context.Validator is changed due to persisted block. As this part has been proven correct, should I revert the newly added UT and resolve this conversation?

Reply:
Are you sure that it's stored that this TX has a `FAULT` state when `FAULT`?

![image](https://user-images.githubusercontent.com/3167973/81665595-2dc6a680-9441-11ea-80df-d9bdc0a41102.png)

Also the default value it's `NONE`

``
  public enum VMState : byte
    {
        NONE = 0,
        HALT = 1,
        FAULT = 2,
        BREAK = 4
    }
``

Reply:
I'm sure. Reasons:

(1) In `clonedSnapshot.Transactions.Add(tx.Hash, state);` the reference of variable `state` is passed rather than the object it self

(2) By `clonedSnapshot.Transactions.Commit();` it can be seen that this transaction state reference has already been added to variable `snapshot`

(3) So in `state.VMState = engine.Execute();` the state recorded in variable `snapshot` is also changed

(4) Then this state will never be touched by other transactions until snapshot is commited.

Reply:
> However this newly added UT brings problems to UT_Consensus, as Consensus.context.Validator is changed due to persisted block. As this part has been proven correct

We can create a simulation test, like the following.

```cs
[TestMethod]
public void Test_Simulate_BlockPersist_SnapshotCommit()
{
     var snapshot = BlockChain.CurrentSnapshot()     
     var cloneSnapshot = snapshot.clone()
      clonedSnapshot.Transactions.Add(tx.Hash, state);
      clonedSnapshot.Transactions.Commit();
      state.VMState = VMState.FAULT
     
     var txState = snapshot.TryGet(tx.Hash)
     txState.VMState.Should().Be(VMState.FAULT)
}
```

Reply:
I tested it and it works, you are right :)

Reply:
Fix the UT and I can approve it